File: app-server/src/test/java/com/github/arhor/simple/expense/tracker/web/security/CreatingUserAuthenticationSuccessHandlerTest.java
package com.github.arhor.simple.expense.tracker.web.security;

import java.io.IOException;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.security.core.Authentication;

import com.github.arhor.simple.expense.tracker.service.UserService;

import static org.mockito.BDDMockito.then;

@ExtendWith(MockitoExtension.class)
class CreatingUserAuthenticationSuccessHandlerTest {

    @Mock
    private UserService userService;

    @Mock
    private HttpServletRequest request;

    @Mock
    private HttpServletResponse response;

    @Mock
    private Authentication authentication;

    @InjectMocks
    private CreatingUserAuthenticationSuccessHandler authenticationSuccessHandler;

    @Test
    void should_call_createNewUserIfNecessary_on_UserService_on_authentication_success()
        throws ServletException, IOException {
        // when
        authenticationSuccessHandler.onAuthenticationSuccess(request, response, authentication);

        // then
        then(userService)
            .should()
            .createNewUserIfNecessary(authentication);
    }
}

File: app-server/src/test/kotlin/com/github/arhor/simple/expense/tracker/web/security/CreatingUserAuthenticationSuccessHandlerTest.kt
package com.github.arhor.simple.expense.tracker.web.security

import com.github.arhor.simple.expense.tracker.service.UserService
import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import io.mockk.impl.annotations.RelaxedMockK
import io.mockk.junit5.MockKExtension
import io.mockk.just
import io.mockk.runs
import io.mockk.verify
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@ExtendWith(MockKExtension::class)
internal class CreatingUserAuthenticationSuccessHandlerTest {

    @MockK
    private lateinit var userService: UserService

    @MockK
    private lateinit var authentication: OAuth2AuthenticationToken

    @RelaxedMockK
    private lateinit var request: HttpServletRequest

    @RelaxedMockK
    private lateinit var response: HttpServletResponse

    @InjectMockKs
    private lateinit var authenticationSuccessHandler: CreatingUserAuthenticationSuccessHandler

    @Test
    fun `should call UserService # createNewUserIfNecessary on authentication success`() {
        // given
        every { userService.createNewUserIfNecessary(any()) } just runs
        every { request.getSession(any()) } answers { callOriginal() }

        // when
        authenticationSuccessHandler.onAuthenticationSuccess(request, response, authentication)

        // then
        verify(exactly = 1) { userService.createNewUserIfNecessary(auth = authentication) }
    }
}
